File: Elvis/quickSort/partitioning.js
let partition = (array) =>{
    if(array.length === 1 || array.length === 0) return array;
    else if(array.length == 2){
        if(array[length - 1] < array [0]){
            let temp = array[0];
            array[0] = array[length - 1];
            array[length - 1] = temp;
        }
        return array;
    }

    let pivot = array[array.length - 1];
    let i = 0, j = array.length - 2;

    while(i < j){
        let rightPointer = array[array.length - j];
        let leftPointer = array[i];
        if(leftPointer < pivot ){
            i++;
        }else{
            if(rightPointer > pivot){
                j--;
            }else{
                array[array.length - j] = leftPointer;
                array[i] = rightPointer;
                i++;
            }
        }
    }
    let temp = array[i];
    array[i] = pivot;
    array[array.length - 1]  = temp


    return array;

}

File: Trash/partitioning.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> partition(vector<int> &arr1)
{
    vector<int> arr = arr1;
    int pivot = arr[arr.size() - 1];
    int leftPointer = 0, rightPointer = arr.size() - 2;
    while (leftPointer < rightPointer)
    {

        while (arr[leftPointer] < pivot)
            leftPointer++;
        while (arr[rightPointer] > pivot)
            rightPointer--;

        if (arr[leftPointer] >= pivot && arr[rightPointer] <= pivot)
        {
            int temp = arr[leftPointer];
            arr[leftPointer] = arr[rightPointer];
            arr[rightPointer] = temp;
            leftPointer++;
            rightPointer--;
        }
    }
    arr[arr.size() - 1] = arr[leftPointer];
    arr[leftPointer] = pivot;
    return arr;
}

int main()
{
    vector<int> arr = {1, 5, 2, 9, 4};
    vector<int> arr1 = partition(arr);
    for(int i = 0; i<arr1.size(); i++) cout<<arr1[i];
    return 0;
}